src/Avalonia.FreeDesktop/Notifications/FreeDesktopNotificationManager.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Avalonia.Controls.Notifications;
+using Tmds.DBus;
+
+namespace Avalonia.FreeDesktop.Notifications
+{
+    public class FreeDesktopNotificationManager : INotificationManager, IDisposable
+    {
+        private const string NotificationsService
+            = "org.freedesktop.Notifications";
+
+        // ReSharper disable once InconsistentNaming
+        private const int DEFAULT_NOTIFICATION_EXPIRATION = -1;
+
+        // ReSharper disable once InconsistentNaming
+        private const int FOREVER_NOTIFICATION_EXPIRATION = 0;
+
+        private static readonly ObjectPath NotificationsPath
+            = new ObjectPath("/org/freedesktop/Notifications");
+
+        private readonly Dictionary<uint, INotification> _notifications;
+
+        private readonly IFreeDesktopNotificationsProxy _proxy;
+        private IDisposable _actionWatcher;
+        private IDisposable _closeNotificationWatcher;
+        private volatile bool _isConnected;
+
+        public FreeDesktopNotificationManager()
+        {
+            _notifications = new Dictionary<uint, INotification>();
+            _proxy = Connection.Session.CreateProxy<IFreeDesktopNotificationsProxy>(
+                NotificationsService,
+                NotificationsPath
+            );
+        }
+
+        public void Dispose()
+        {
+            _actionWatcher?.Dispose();
+            _closeNotificationWatcher?.Dispose();
+        }
+
+        public async ValueTask ShowAsync(INotification notification)
+        {
+            if (!_isConnected)
+            {
+                await Connect();
+            }
+
+            var id = await _proxy.NotifyAsync(
+                Application.Current.Name,
+                0,
+                //TODO: Impl app icon
+                string.Empty,
+                notification.Title,
+                notification.Message,
+                Array.Empty<string>(),
+                GetHintsFromNotification(notification),
+                GetExpirationInMsFromNotification(notification)
+            ).ConfigureAwait(false);
+
+            _notifications[id] = notification;
+        }
+
+        private async ValueTask Connect()
+        {
+            _isConnected = await
```

Comment:
Note that the app might be started without a session bus or without `org.freedesktop.Notifications` service on the session bus. That's the case for VNC-powered sessions, WSL, embedded devices and other scenarios without a "full" desktop environment. We should gracefully fall back to some managed implementation in that case.

Use `DBusHelper.Connection` which will be just `null` if no session DBus is available like [we are doing in our dbusmenu implementation](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.FreeDesktop/DBusMenuExporter.cs#L20). Then try to create a proxy and see if that throws. That check should be done before registering `FreeDesktopNotificationManager` in the service locator, so the rest of the code base won't be confused by dysfunctional notification manager. It would also be a good place to call `GetCapabilitiesAsync`.

Reply:
> It would also be a good place to call GetCapabilitiesAsync.

Which capabilties would you check for? At first glance the features we rely on should be working in all server implementations.

Reply:
Currently we're only invoking `DBusHelper.TryInitialize()` when the `UseDBusMenu` is enabled. We should probably be doing this unconditionally now.

Reply:
>Which Window should it be installed on though

The initial idea was that user code is supposed to get the notification manager from the window itself rather than using service locator. A single app might be using toplevels represented by different windowing systems. I. e. it's planned to have a "remote" mode for devtools when they are using our previewer remote widget infrastructure to render themselves to a remote machine (i. e. you will be able to connect to devtools of an application running on your android phone). So the general policy is to remove global variables when possible.

So there shouldn't be a globally accessible instance of the notification manager, only one accessed via ITopLevelImpl in some manner.

Reply:
> So there shouldn't be a globally accessible instance of the notification manager, only one accessed via ITopLevelImpl in some manner.

Allright, but that seems wastly out of scope for this PR. For now the service will just return `null` when it fails to initialize. I guess the users can account for that in their code and fire up the `WindowNotificationManager` instead.

Reply:
> that seems vastly out of scope for this PR

the system notification manager should still be accessed via the TopLevel though.

Reply:
Wait... So instead of registering the notification manager (on all platforms) through `AvaloniaLocator` I should just make it a property on the `TopLevel`?

Reply:
Yes, probably an attached one. The general idea is that services should be resolved via toplevels (somewhat like in Android where every damn thing needs a `Context`)

Reply:
Also, note how it works with global menu - toplevel impl implements an extra interface where it announces that it could potentially provide some service.

Reply:
> toplevel impl implements an extra interface

I can't seem to find the interface you're referring to. What is it called?

Reply:
https://github.com/AvaloniaUI/Avalonia/blob/e80ddf5c6591bff9e936fa3c8655e0daff9fd39b/src/Avalonia.Controls/Platform/ITopLevelNativeMenuExporter.cs#L14